### PR TITLE
riscv/fpu: Implement correct lazy-FPU functionality

### PR DESCRIPTION
**Pull request by pussuw**
## Summary
This implements / restores lazy-FPU functionality which was lost after the user context was moved into the stack. It moves FPU context away from the stack into a separate FPU save area in the TCB. This area is not volatile / prone to the stack moving, thus it is safe to use for lazy FPU saving / restoring.
## Impact
Architecturally quite significant, but brings a nice performance boost and also fixes a nasty bug in the current FPU save/restore.

Some statistics; https://github.com/apache/nuttx/pull/9103#discussion_r1214096917
2.7% decrease in CPU load is well worth the change.
## Testing
icicle:nsh + icicle:knsh

NOTE:
This (https://github.com/apache/nuttx/pull/9485) change is required in conjunction for CONFIG_BUILD_KERNEL, otherwise there is a risk of system crash when saving FPU state. It happens when a temporary address environment switch is active and the process's own address environment is not available (happens during posix_spawn() for example)

### CODE REVIEW

File: arch/risc-v/src/common/riscv_fpu.S [hunk 1]
```diff
@@ -57,18 +61,95 @@
...
+
+    .type         riscv_savefpu, function
+
+riscv_savefpu:
+
+  REGLOAD    t0, REG_INT_CTX(a0)
+  li         t1, MSTATUS_FS
+  and        t2, t0, t1
+  li         t1, MSTATUS_FS_DIRTY
+  bne        t2, t1, 1f
```
Comment (pussuw) [inline arch/risc-v/src/common/riscv_fpu.S:106]:
```
Just a reminder to self:

#ifdef CONFIG_ARCH_RISCV_LAZYFPU
bne t2, t1, 1f
#else
blt t2, t1, 1f
#endif
```

**Comment by pussuw** [comment]
Some kind of wget error / similar.

====================================================================================
Configuration/Tool: icicle/rpmsg-ch1
------------------------------------------------------------------------------------
  Cleaning...
  Configuring...
  Building NuttX...
  End-of-central-directory signature not found.  Either this file is not
  a zipfile, or it constitutes one disk of a multi-part archive.  In the
  latter case the central directory and zipfile comment will be found on
  the last disk(s) of this archive.
unzip:  cannot find zipfile directory in one of libmetal.zip or
        libmetal.zip.zip, and cannot find libmetal.zip.ZIP, period.
make[1]: *** [libmetal.defs:76: libmetal.zip] Error 9
make[1]: Target 'context' not remade because of errors.
make: *** [tools/Unix.mk:443: openamp/.context] Error 2
make: Target 'all' not remade because of errors.
/github/workspace/sources/nuttx/tools/testbuild.sh: line 338: /github/workspace/sources/nuttx/../nuttx/nuttx.manifest: No such file or directory
  Normalize icicle/rpmsg-ch1

**Comment by pussuw** [comment]
@xiaoxiang781216 the option to disable lazy FPU is now implemented also, tested on my app that both lazy and non-lazy modes work

**Comment by xiaoxiang781216** [comment]
I will review in the weekend.

**Comment by jerpelea** [comment]
please rebase

**Comment by xiaoxiang781216** [comment]
@pussuw could you rebase the change?

**Comment by pussuw** [comment]
> @pussuw could you rebase the change?

Yes I will do it tomorrow

File: arch/risc-v/include/irq.h [hunk 2]
```diff
@@ -204,50 +204,56 @@
...
+#  define REG_F26_NDX       (FPU_REG_SIZE * 26)
+#  define REG_F27_NDX       (FPU_REG_SIZE * 27)
+#  define REG_F28_NDX       (FPU_REG_SIZE * 28)
+#  define REG_F29_NDX       (FPU_REG_SIZE * 29)
+#  define REG_F30_NDX       (FPU_REG_SIZE * 30)
+#  define REG_F31_NDX       (FPU_REG_SIZE * 31)
+#  define REG_FCSR_NDX      (FPU_REG_SIZE * 32)
+
+#  define FPU_XCPT_NREGS    (33)
+#  define FPU_XCPT_REGS     (FPU_REG_SIZE * FPU_XCPT_NREGS)
```
Comment (xiaoxiang781216) [inline arch/risc-v/include/irq.h:242]: change to:
```
#  define FPU_XCPT_REGS     (FPU_REG_SIZE * 33)
```
and remove line 241

File: arch/risc-v/include/irq.h [hunk 3]
```diff
@@ -204,50 +204,56 @@
...
 #  define FPU_REG_FULL_SIZE (INT_REG_SIZE * FPU_REG_SIZE)
+#  define FPU_XCPT_SIZE     (FPU_REG_FULL_SIZE * FPU_XCPT_NREGS)
```
Comment (xiaoxiang781216) [inline arch/risc-v/include/irq.h:244]: change to:
```
define FPU_XCPT_SIZE     (FPU_XCPT_REGS * FPU_XCPT_REGS)
```
and remove FPU_REG_FULL_SIZE.

File: arch/risc-v/src/common/riscv_macros.S [hunk 4]
```diff
@@ -101,11 +101,6 @@
 .macro riscv_savefpu in
 
 #ifdef CONFIG_ARCH_FPU
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/common/riscv_macros.S:103]: could remove "#ifdef" check

File: arch/risc-v/src/common/riscv_macros.S [hunk 5]
```diff
@@ -217,11 +206,6 @@
 .macro riscv_loadfpu out
 
 #ifdef CONFIG_ARCH_FPU
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/common/riscv_macros.S:208]: could remove "#ifdef" check too

File: arch/risc-v/src/common/riscv_fpu.S [hunk 6]
```diff
@@ -57,18 +61,95 @@
...
+ *
+ ****************************************************************************/
+
+    .type        riscv_restorefpu, function
+
+riscv_restorefpu:
+
+  REGLOAD      t0, REG_INT_CTX(a0)
+  li           t1, MSTATUS_FS
+  and          t2, t0, t1
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/common/riscv_fpu.S:150]: could change t2 to t0

File: arch/risc-v/include/irq.h (hunk 3 above)
Comment (pussuw) [reply arch/risc-v/include/irq.h:244]: FPU_REG_FULL_SIZE is used by riscv_vfork.h:
https://github.com/apache/nuttx/blob/e9738cd906710cf3b4db3816a8e6f5810bcd6772/arch/risc-v/src/common/riscv_vfork.h#L92-L105

File: arch/risc-v/src/common/riscv_macros.S (hunk 4 above)
Comment (pussuw) [reply arch/risc-v/src/common/riscv_macros.S:103]: Sure, I'll remove the other one too

File: arch/risc-v/src/common/riscv_fpu.S (hunk 6 above)
Comment (pussuw) [reply arch/risc-v/src/common/riscv_fpu.S:150]: True, but let's keep it as it does not change anything

File: arch/risc-v/include/irq.h (hunk 3 above)
Comment (xiaoxiang781216) [reply arch/risc-v/include/irq.h]: Ok, let's keep it.

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by masayuki2009** [comment]
@pussuw 
Hmm, I noticed that ostest with maix-bit:smp (board) crashes now.

```
user_main: setvbuf test
setvbuf_test: Test NO buffering
setvbuf_test: Using NO buffering
setvbuf_test: Test default FULL buffering
setvbuf_test: Using default FULL buffering
setvbuf_test: Test FULL buffering, buffer size 64
setvbuf_test: Using FULL buffering, buffer size 64
setvbuf_test: Test FULL buffering, pre-allocated buffer
setvbuf_test: Using FULL buffering, pre-allocated buffer
setvbuf_test: Test LINE buffering, buffer size 64
setvbuf_test: Using LINE buffering, buffer size 64
setvbuf_test: Test FULL buffering, pre-allocated buffer
setvbuf_test: Using FULL buffering, pre-allocated buffer

End of test memory usage:
VARIABLE  BEFORE   AFTER
======== ======== ========
arena      1fb990   1fb990
ordblks         4        4
mxordblk   1f4020   1f4020
uordblks     7100     7100
fordblks   1f4890   1f4890

user_main: /dev/null test
dev_null: Read 0 bytes from /dev/null
dev_null: Wrote 1024 bytes to /dev/null

End of test [CPU1] riscv_exception: EXCEPTION: Illegal instruction. MCAUSE: 0000000000000002, EPC: 0000000080016bf8, MTVAL: 00000037ec3f44d4
[CPU1] riscv_exception: PANIC!!! Exception = 0000000000000002
[CPU1] _assert: Current Version: NuttX  12.0.0 425cc89989 Jun 19 2023 22:23:28 risc-v
[CPU1] _assert: Assertion failed panic: at file: :0 task(CPU1): FPU#1 0x80016b9c
[CPU1] up_dump_register: EPC: 0000000080016bf8
[CPU1] up_dump_register: A0: 0000000000000000 A1: 0000000000000001 A2: 0000000000000000 A3: 0000000080403d58
[CPU1] up_dump_register: A4: 0000000000000002 A5: 0000000000000440 A6: 0000000000000000 A7: 0000000000000000
[CPU1] up_dump_register: T0: 0000000000000000 T1: 0000000000000000 T2: 0000000000000000 T3: 0000000000000000
[CPU1] up_dump_register: T4: 0000000000000000 T5: 0000000000000000 T6: 0000000000000000
[CPU1] up_dump_register: S0: 0000000000000000 S1: 0000000000000001 S2: 0000000000000000 S3: 0000000000000000
[CPU1] up_dump_register: S4: 0000000000000000 S5: 0000000080403248 S6: 0000000000000000 S7: 0000000000000000
[CPU1] up_dump_register: S8: 0000000000000000 S9: 0000000000000000 S10: 0000000000000000 S11: 0000000000000000
[CPU1] up_dump_register: SP: 000000008040c720 FP: 0000000000000000 TP: 0000000000000000 RA: 0000000080016bea
[CPU1] dump_stack: User Stack:
[CPU1] dump_stack:   base: 0x8040c030
[CPU1] dump_stack:   size: 00001968
[CPU1] dump_stack:     sp: 0x8040c720
[CPU1] stack_dump: 0x8040c720: deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef
[CPU1] stack_dump: 0x8040c740: deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef deadbeef
[CPU1] stack_dump: 0x8040c760: deadbeef deadbeef deadbeef deadbeef 00000000 00000000 00000000 00000000
[CPU1] stack_dump: 0x8040c780: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
[CPU1] stack_dump: 0x8040c7a0: 80016b9c 00000000 80006c8e 00000000 8040c010 00000000 00000001 00000000
[CPU1] stack_dump: 0x8040c7c0: 00000000 00000000 8000364c 00000000 00000000 00000000 00000000 00000000
[CPU1] show_tasks:    PID GROUP   CPU PRI POLICY   TYPE    NPX STATE   EVENT      SIGMASK          STACKBASE  STACKSIZE      USED   FILLED    COMMAND
[CPU1] show_tasks:   ----   ---  ---- --- -------- ------- --- ------- ---------- -------- 0x80400240      2048      1384    67.5%    irq
[CPU1] dump_task:       0     0     0   0 FIFO     Kthread N-- Running            0000000000000000 0x80403ea0      2000      1136    56.8%    CPU0 IDLE
[CPU1] dump_task:       1     1     1   0 FIFO     Kthread N-- Assigned           0000000000000000 0x80401270      2000      1104    55.2%    CPU1 IDLE
[CPU1] dump_task:       2     2     0 100 RR       Task    --- Waiting Semaphore  0000000000000000 0x80406670      2992      2176    72.7%    nsh_main
[CPU1] dump_task:      39    39     0 100 RR       Task    --- Waiting Semaphore  0000000000000000 0x8040aa20      1968      1456    73.9%    ostest
[CPU1] dump_task:      40    40     0 100 RR       Task    --- Waiting Semaphore  0000000000000000 0x80407990      8064      1520    18.8%    ostest Arg1 Arg2 Arg3 Arg4
[CPU1] dump_task:      41    41     1 100 RR       Task    --- Running            0000000000000000 0x8040c030      1968       720    36.5%    FPU#1
```

**Comment by masayuki2009** [comment]
>@pussuw
>Hmm, I noticed that ostest with maix-bit:smp (board) crashes now.

Also, rv-virt:ksmp64 crashes on boot now.

```
/home/ishikawa/opensource/QEMU/qemu-7.1/build/qemu-system-riscv64 -semihosting -nographic -cpu rv64 -smp 8 -M virt,aclint=on -bios none -kernel nuttx
[    0.100000] [CPU1] riscv_exception: EXCEPTION: Load page fault. MCAUSE: 000000000000000d, EPC: 0000000080004c14, MTVAL: 00000000c02002f0
[    0.110000] [CPU1] riscv_exception: PANIC!!! Exception = 000000000000000d
[    0.110000] [CPU1] _assert: Current Version: NuttX  12.0.0 425cc89989 Jun 19 2023 22:24:42 risc-v
[    0.110000] [CPU1] _assert: Assertion failed panic: at file: :0 task(CPU1): /system/bin/init 0xc000004a
[    0.110000] [CPU1] up_dump_register: EPC: 0000000080004c14
[    0.110000] [CPU1] up_dump_register: A0: 000000008020d6f0 A1: 0000000000000101 A2: 0000000000000000 A3: 0000000000000000
[    0.110000] [CPU1] up_dump_register: A4: 00000000c02002d0 A5: 000000008020d6f0 A6: 0000000000000000 A7: 0000000000000000
[    0.120000] [CPU1] up_dump_register: T0: 0000000000000000 T1: 0000000000000000 T2: 0000000000000000 T3: 0000000000000000
[    0.120000] [CPU1] up_dump_register: T4: 0000000000000000 T5: 0000000000000000 T6: 0000000000000000
[    0.120000] [CPU1] up_dump_register: S0: 0000000000000000 S1: 0000000000000000 S2: 0000000000000000 S3: 0000000000000000
[    0.120000] [CPU1] up_dump_register: S4: 0000000000000000 S5: 0000000000000000 S6: 0000000000000000 S7: 0000000000000000
[    0.130000] [CPU1] up_dump_register: S8: 0000000000000000 S9: 0000000000000000 S10: 0000000000000000 S11: 0000000000000000
[    0.130000] [CPU1] up_dump_register: SP: 000000008020f400 FP: 0000000000000000 TP: 0000000000000000 RA: 0000000080004c0a
[    0.130000] [CPU1] dump_stack: Kernel Stack:
[    0.130000] [CPU1] dump_stack:   base: 0x8020e810
[    0.130000] [CPU1] dump_stack:   size: 00003072
[    0.130000] [CPU1] dump_stack:     sp: 0x8020f400
[    0.130000] [CPU1] show_tasks:    PID GROUP   CPU PRI POLICY   TYPE    NPX STATE   EVENT      SIGMASK          STACKBASE  STACKSIZE      USED   FILLED    COMMAND
[    0.140000] [CPU1] show_tasks:   ----   ---  ---- --- -------- ------- --- ------- ---------- -------- 0x802002b0      2048       104     5.0%    irq
[    0.140000] [CPU1] dump_task:       0     0     0   0 FIFO     Kthread N-- Running            0000000000000000 0x8020a030      3024      1592    52.6%    CPU0 IDLE
[    0.140000] [CPU1] dump_task:       1     1     1   0 FIFO     Kthread N-- Assigned           0000000000000000 0x80202310      3024       584    19.3%    CPU1 IDLE
[    0.140000] [CPU1] dump_task:       2     2     2   0 FIFO     Kthread N-- Running            0000000000000000 0x80202f10      3024       584    19.3%    CPU2 IDLE
[    0.150000] [CPU1] dump_task:       3     3     3   0 FIFO     Kthread N-- Running            0000000000000000 0x80203b10      3024       584    19.3%    CPU3 IDLE
[    0.150000] [CPU1] dump_task:       4     4     0 100 RR       Kthread --- Waiting Semaphore  0000000000000000 0x8020e060      1952       408    20.9%    lpwork 0x802056c8
[    0.150000] [CPU1] dump_task:       5     5     1 100 RR       Task    --- Running            0000000000000000 0xc0202060      2976         0     0.0%    /system/bin/init
```

**Comment by pussuw** [comment]
> @pussuw May I revert this PR since this PR is unstable?

Yes, I will run ksmp64 tomorrow to see what is wrong. Both are smp targets, I did not test smp as I don't own smp capable boards.

**Comment by pussuw** [comment]
@masayuki2009 did you notice anything wrong with non-smp targets?

**Comment by masayuki2009** [comment]
> @masayuki2009 did you notice anything wrong with non-smp targets?

@pussuw 
I tried both maix-bit:smp (board) and rv-virt:ksmp64 with CONFIG_SMP_NCPUS=1 and found that the issues do not happen.

**Comment by pussuw** [comment]
> > @masayuki2009 did you notice anything wrong with non-smp targets?
> 
> @pussuw I tried both maix-bit:smp (board) and rv-virt:ksmp64 (QEMU) with CONFIG_SMP_NCPUS=1 and found that the issues do not happen.

Thanks!

So it does seem smp might be the problem. Kind of good and kind of bad. Bad because I don't know anything about NuttX smp :(

**Pull request merged**